### ISSUE TITLE
sam build stop working on sam-cli 1.6.2 building ruby 2.5 project on CircleCI

### ISSUE THREAD

**Issue by wildodecisiv** [aws/aws-sam-cli#2292]
### Description

Sam build started throwing an exception when trying to `bundle install called by sam build` the gems of my ruby lambda project.
I have checked my historical builds on CircleCI and I found that `sam build` worked perfectly with `sam-cli 1.4.0` and it started failing with the new `sam-client 1.6.2.`

### Steps to reproduce
1. On CircleCI set docker image to: lambci/lambda:build-ruby2.5
**Note:** if you set to `lambci/lambda:20201004-build-ruby2.5` it works as expected because it uses `sam-cli 1.4.0`, but the last docker image is using  `sam 1.6.2`
2. Run "sam build -t template.yaml --debug" of a ruby with lambda and layers project

### Observed result

```
.....
.....
Found workflow 'RubyBundlerBuilder' to support capabilities 'Capability(language='ruby', dependency_manager='bundler', application_framework=None)'
Running workflow 'RubyBundlerBuilder'
Running RubyBundlerBuilder:CopySource
RubyBundlerBuilder:CopySource succeeded
Running RubyBundlerBuilder:RubyBundle
Running bundle install in /tmp/tmp6kibmw5f
executing Bundler: ['bundle', 'install', '--without', 'development', 'test']
RubyBundlerBuilder:RubyBundle failed
Traceback (most recent call last):
  File "/usr/local/pipx/venvs/aws-sam-cli/lib64/python3.6/site-packages/aws_lambda_builders/workflows/ruby_bundler/actions.py", line 31, in execute
    self.subprocess_bundler.run(["install", "--without", "development", "test"], cwd=self.source_dir)
  File "/usr/local/pipx/venvs/aws-sam-cli/lib64/python3.6/site-packages/aws_lambda_builders/workflows/ruby_bundler/bundler.py", line 55, in run
    raise BundlerExecutionError(message=out.decode("utf8").strip())
aws_lambda_builders.workflows.ruby_bundler.bundler.BundlerExecutionError: Bundler Failed: 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/pipx/venvs/aws-sam-cli/lib64/python3.6/site-packages/aws_lambda_builders/workflow.py", line 269, in run
    action.execute()
  File "/usr/local/pipx/venvs/aws-sam-cli/lib64/python3.6/site-packages/aws_lambda_builders/workflows/ruby_bundler/actions.py", line 33, in execute
    raise ActionFailedError(str(ex))
aws_lambda_builders.actions.ActionFailedError: Bundler Failed: 

Build Failed
Sending Telemetry: {'metrics': [{'commandRun': {'awsProfileProvided': False, 'debugFlagProvided': True, 'region': '', 'commandName': 'sam build', 'duration': 390, 'exitReason': 'WorkflowFailedError', 'exitCode': 1, 'requestId': '365b5106-d1b5-449e-8fed-2d6e9c38df14', 'installationId': 'a7397019-1cc4-4735-910e-118275815766', 'sessionId': '6fb2da56-c041-4406-a696-ef36457dbd48', 'executionEnvironment': 'CLI', 'pyversion': '3.6.12', 'samcliVersion': '1.6.2'}}]}
Telemetry response: 200
Error: RubyBundlerBuilder:RubyBundle - Bundler Failed: 

Exited with code exit status 1
CircleCI received exit code 1
```

### Expected result

I'm expecting a build successfull message

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS:CircleCI docker image lambci/lambda:build-ruby2.5
2. `sam --version`: 1.6.2

`Add --debug flag to command you are running`

**Comment by mndeveci** [comment]
Hi @wildodecisiv ,

I tried to replicate the issue with a simple ruby 2.5 hello world application but it is build successfully.
I also tried with this example repo, https://github.com/jkahn117/aws-lambda-ruby-layers (which contains a layer) but that one also built successfully.

Can you give an example that we can re-produce?

Thanks

**Comment by wildodecisiv** [comment]
Hi @mndeveci , thanks for try to fix the issue

I created a project in order to help you to reproduce the issue, I hope this can help you.
Github: https://github.com/wildodecisiv/ruby-project-example

It uses docker-compose in order to easy try to build the project using the sam client.
Please take a look at the readme file, I left the steps to build successfully using sam 1.4.0 as well as to get the error
using sam 1.6.3

Thank you

**Comment by mndeveci** [comment]
Hi @wildodecisiv ,

Thanks for the detailed example!

With the version 1.6.2, we introduced de-duplicated builds. Before that each function was building into `.aws-sam/build/{function_name}` folder, but with this change, each function will be build in a temporary folder, and then copy artifacts for the `.aws-sam/build/{function_name}`. This behaviour change might be affecting your example.

I tried to copy Gemfile into `app/functions/ruby_example` and re-run the failing build, and it was successful with the latest image. This might be used as workaround for now.

Can you also provide an example which won't use the docker image? Like an example where I can run `sam build` directly from command line?

Thanks!

**Comment by wildodecisiv** [comment]
Thanks for the update @mndeveci 
I don't have an example right now of a project without docker image but I think that you can run  `sam build` directly in your terminal that will produce the same result (if you remove gem pg from the Gemfile), I just provided this example with docker-compose/docker because I use native gems like `pg` that require your OS to have installed postgresql native dependencies.

I appreciate the workaround  you suggested to do related to copy a Gemfile into `app/functions/ruby_example`, but as I'm using a lambda layer to handle the Gemfile as well as other common ruby codes I think that the Gemfile into that folder is only to build successfully and it does not add a value to the function in my opinion. 
The workaround that I found was to use the docker image with sam client 1.4.0 where it worked successfully.

My concern is that  I have 5 projects on Production pretty much with the same structure I used for this ruby project example and each one started to fails at the build step on CircleCI after the automatic sam cli upgrade because we are pulling always the last docker image for ruby2.5 that now comes with sam client 1.6.2
I had to update the code for each project in order to use the docker image with sam-cli 1.4.0
I dont see any update in the sam documentation as well as ruby documents using lamba layers where mentioned that each function must have a Gemfile in order to build successfully the app, maybe the doc needs to be updated? with the new version?

I appreciate all the work here! if there is a workaround to fix this maybe it is not a critical issue (if you have 5 projects but if someone has 100 projects will become a headache to update all projects to include a Gemfile to the function folder)
In my opinion is an important issue, maybe it is wrong practice to not have a Gemfile for each function (if you want to handle gems using layers) that could rise an exemption for future versions of the sam client but it should be displayed just as a "Warning" in my opinion that will be mandatory for future versions in order to have time to fix it.

Thanks again and I leave to your discretion to close the ticket or not knowing that we can do a workaround. :)

**Comment by mndeveci** [comment]
Hi @wildodecisiv ,

Sorry I think I didn't myself clear that, the workaround that I propose was to unblock you and continue your development, it wasn't suppose to be a permanent solution.

And the reason I asked for an example that we can run locally is for validating the fix that we will be working on. With that example we can't validate any changes because it is using some 3rd party image that we don't support, and have no control. Having an example that we can just run `sam build` locally would make testing much more easier.

I labeled this issue as bug and we will be working on a fix soon.

**Comment by mndeveci** [comment]
The fix has been released with [SAM CLI v1.7.0](https://github.com/aws/aws-sam-cli/releases/tag/v1.7.0).
